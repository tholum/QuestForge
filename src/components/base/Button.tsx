import * as React from "react"
import { Slot } from "@radix-ui/react-slot"
import { cva, type VariantProps } from "class-variance-authority"
import { Loader2 } from "lucide-react"

import { cn } from "@/lib/utils"

const buttonVariants = cva(
  "inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all duration-200 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive focus:scale-[0.98] active:scale-[0.96] select-none",
  {
    variants: {
      variant: {
        default:
          "bg-primary text-primary-foreground shadow-xs hover:bg-primary/90 hover:shadow-sm",
        destructive:
          "bg-destructive text-white shadow-xs hover:bg-destructive/90 hover:shadow-sm focus-visible:ring-destructive/20 dark:focus-visible:ring-destructive/40 dark:bg-destructive/60",
        outline:
          "border bg-background shadow-xs hover:bg-accent hover:text-accent-foreground hover:shadow-sm dark:bg-input/30 dark:border-input dark:hover:bg-input/50",
        secondary:
          "bg-secondary text-secondary-foreground shadow-xs hover:bg-secondary/80 hover:shadow-sm",
        ghost:
          "hover:bg-accent hover:text-accent-foreground dark:hover:bg-accent/50",
        link: "text-primary underline-offset-4 hover:underline",
        success:
          "bg-green-600 text-white shadow-xs hover:bg-green-700 hover:shadow-sm focus-visible:ring-green-500/20 dark:bg-green-500/90 dark:hover:bg-green-600",
        warning:
          "bg-yellow-600 text-white shadow-xs hover:bg-yellow-700 hover:shadow-sm focus-visible:ring-yellow-500/20 dark:bg-yellow-500/90 dark:hover:bg-yellow-600",
      },
      size: {
        xs: "h-6 text-xs px-2 gap-1 has-[>svg]:px-1.5",
        sm: "h-8 text-sm px-3 gap-1.5 has-[>svg]:px-2.5",
        default: "h-9 px-4 py-2 has-[>svg]:px-3",
        lg: "h-10 px-6 has-[>svg]:px-4",
        xl: "h-12 px-8 text-base has-[>svg]:px-6",
        icon: "size-9",
        "icon-sm": "size-8",
        "icon-lg": "size-10",
        "icon-xl": "size-12",
      },
      rounded: {
        default: "rounded-md",
        sm: "rounded-sm",
        lg: "rounded-lg",
        xl: "rounded-xl",
        full: "rounded-full",
        none: "rounded-none",
      },
    },
    defaultVariants: {
      variant: "default",
      size: "default",
      rounded: "default",
    },
  }
)

export interface IButtonProps extends 
  Omit<React.ComponentProps<"button">, "children">,
  VariantProps<typeof buttonVariants> {
  asChild?: boolean
  loading?: boolean
  loadingText?: string
  leftIcon?: React.ReactNode
  rightIcon?: React.ReactNode
  children?: React.ReactNode
  fullWidth?: boolean
  tooltip?: string
  badge?: number | string
}

const Button = React.forwardRef<HTMLButtonElement, IButtonProps>(
  ({
    className,
    variant,
    size,
    rounded,
    asChild = false,
    loading = false,
    loadingText,
    leftIcon,
    rightIcon,
    children,
    fullWidth = false,
    tooltip,
    badge,
    disabled,
    ...props
  }, ref) => {
    const Comp = asChild ? Slot : "button"
    
    const isDisabled = disabled || loading

    const buttonContent = (
      <>
        {loading ? (
          <Loader2 className="animate-spin" />
        ) : (
          leftIcon
        )}
        
        <span className="flex-1">
          {loading && loadingText ? loadingText : children}
        </span>
        
        {!loading && rightIcon}
        
        {badge !== undefined && (
          <span className="ml-1 inline-flex items-center justify-center min-w-5 h-5 text-xs font-medium bg-red-500 text-white rounded-full px-1">
            {badge}
          </span>
        )}
      </>
    )

    return (
      <Comp
        ref={ref}
        data-slot="button"
        data-loading={loading}
        title={tooltip}
        disabled={isDisabled}
        className={cn(
          buttonVariants({ variant, size, rounded }),
          fullWidth && "w-full",
          className
        )}
        {...props}
      >
        {buttonContent}
      </Comp>
    )
  }
)

Button.displayName = "Button"

export { Button, buttonVariants }