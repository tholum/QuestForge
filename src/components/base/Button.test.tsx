import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { describe, it, expect, vi } from 'vitest'
import { Button } from './Button'
import { Plus, ArrowRight } from 'lucide-react'

describe('Button', () => {
  it('renders button with text', () => {
    render(<Button>Click me</Button>)
    expect(screen.getByRole('button')).toHaveTextContent('Click me')
  })

  it('handles click events', async () => {
    const handleClick = vi.fn()
    const user = userEvent.setup()
    
    render(<Button onClick={handleClick}>Click me</Button>)
    
    await user.click(screen.getByRole('button'))
    expect(handleClick).toHaveBeenCalledTimes(1)
  })

  it('applies variant classes correctly', () => {
    const { rerender } = render(<Button variant="success">Button</Button>)
    expect(screen.getByRole('button')).toHaveClass('bg-green-600')
    
    rerender(<Button variant="destructive">Button</Button>)
    expect(screen.getByRole('button')).toHaveClass('bg-destructive')
  })

  it('applies size classes correctly', () => {
    const { rerender } = render(<Button size="sm">Button</Button>)
    expect(screen.getByRole('button')).toHaveClass('h-8')
    
    rerender(<Button size="lg">Button</Button>)
    expect(screen.getByRole('button')).toHaveClass('h-10')
  })

  it('shows loading state correctly', () => {
    render(<Button loading>Save</Button>)
    
    // Should show spinner (Loader2 from lucide-react)
    const button = screen.getByRole('button')
    expect(button.querySelector('svg')).toBeInTheDocument()
    expect(button).toHaveAttribute('data-loading', 'true')
    
    // Button should be disabled when loading
    expect(button).toBeDisabled()
  })

  it('shows loading text when provided', () => {
    render(<Button loading loadingText="Saving...">Save</Button>)
    expect(screen.getByRole('button')).toHaveTextContent('Saving...')
  })

  it('renders with left icon', () => {
    render(
      <Button leftIcon={<Plus data-testid="plus-icon" />}>
        Add Item
      </Button>
    )
    
    expect(screen.getByTestId('plus-icon')).toBeInTheDocument()
    expect(screen.getByRole('button')).toHaveTextContent('Add Item')
  })

  it('renders with right icon', () => {
    render(
      <Button rightIcon={<ArrowRight data-testid="arrow-icon" />}>
        Continue
      </Button>
    )
    
    expect(screen.getByTestId('arrow-icon')).toBeInTheDocument()
    expect(screen.getByRole('button')).toHaveTextContent('Continue')
  })

  it('renders with both left and right icons', () => {
    render(
      <Button 
        leftIcon={<Plus data-testid="plus-icon" />}
        rightIcon={<ArrowRight data-testid="arrow-icon" />}
      >
        Add and Continue
      </Button>
    )
    
    expect(screen.getByTestId('plus-icon')).toBeInTheDocument()
    expect(screen.getByTestId('arrow-icon')).toBeInTheDocument()
  })

  it('shows badge when provided', () => {
    render(<Button badge={5}>Messages</Button>)
    expect(screen.getByText('5')).toBeInTheDocument()
    expect(screen.getByRole('button')).toHaveTextContent('Messages')
  })

  it('shows string badge', () => {
    render(<Button badge="NEW">Features</Button>)
    expect(screen.getByText('NEW')).toBeInTheDocument()
  })

  it('applies fullWidth class', () => {
    render(<Button fullWidth>Full Width</Button>)
    expect(screen.getByRole('button')).toHaveClass('w-full')
  })

  it('is disabled when disabled prop is true', () => {
    render(<Button disabled>Disabled</Button>)
    expect(screen.getByRole('button')).toBeDisabled()
  })

  it('is disabled when loading is true', () => {
    render(<Button loading>Loading</Button>)
    expect(screen.getByRole('button')).toBeDisabled()
  })

  it('sets tooltip title attribute', () => {
    render(<Button tooltip="This is a tooltip">Button</Button>)
    expect(screen.getByRole('button')).toHaveAttribute('title', 'This is a tooltip')
  })

  it('applies custom className', () => {
    render(<Button className="custom-class">Button</Button>)
    expect(screen.getByRole('button')).toHaveClass('custom-class')
  })

  it('applies rounded variant classes', () => {
    const { rerender } = render(<Button rounded="full">Button</Button>)
    expect(screen.getByRole('button')).toHaveClass('rounded-full')
    
    rerender(<Button rounded="none">Button</Button>)
    expect(screen.getByRole('button')).toHaveClass('rounded-none')
  })

  it('renders as child element when asChild is true', () => {
    render(
      <Button asChild>
        <a href="/test">Link Button</a>
      </Button>
    )
    
    const link = screen.getByRole('link')
    expect(link).toHaveAttribute('href', '/test')
    expect(link).toHaveTextContent('Link Button')
  })

  // Accessibility tests
  it('has proper accessibility attributes', () => {
    render(<Button>Accessible Button</Button>)
    const button = screen.getByRole('button')
    
    expect(button).toHaveAttribute('data-slot', 'button')
  })

  it('maintains focus management', async () => {
    const user = userEvent.setup()
    render(<Button>Focusable Button</Button>)
    
    const button = screen.getByRole('button')
    await user.tab()
    
    expect(button).toHaveFocus()
  })

  it('handles keyboard interactions', async () => {
    const handleClick = vi.fn()
    const user = userEvent.setup()
    
    render(<Button onClick={handleClick}>Keyboard Button</Button>)
    
    const button = screen.getByRole('button')
    button.focus()
    
    await user.keyboard('{Enter}')
    expect(handleClick).toHaveBeenCalledTimes(1)
    
    await user.keyboard(' ')
    expect(handleClick).toHaveBeenCalledTimes(2)
  })

  // Snapshot test
  it('matches snapshot', () => {
    const { container } = render(
      <Button 
        variant="success" 
        size="lg" 
        leftIcon={<Plus />}
        badge={3}
      >
        Test Button
      </Button>
    )
    expect(container.firstChild).toMatchSnapshot()
  })
})