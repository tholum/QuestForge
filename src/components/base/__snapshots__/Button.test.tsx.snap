// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > matches snapshot 1`] = `
<button
  class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-all duration-200 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive focus:scale-[0.98] active:scale-[0.96] select-none bg-green-600 text-white shadow-xs hover:bg-green-700 hover:shadow-sm focus-visible:ring-green-500/20 dark:bg-green-500/90 dark:hover:bg-green-600 h-10 px-6 has-[>svg]:px-4 rounded-md"
  data-loading="false"
  data-slot="button"
>
  <svg
    aria-hidden="true"
    class="lucide lucide-plus"
    fill="none"
    height="24"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M5 12h14"
    />
    <path
      d="M12 5v14"
    />
  </svg>
  <span
    class="flex-1"
  >
    Test Button
  </span>
  <span
    class="ml-1 inline-flex items-center justify-center min-w-5 h-5 text-xs font-medium bg-red-500 text-white rounded-full px-1"
  >
    3
  </span>
</button>
`;
